### PR TITLE
fix out of place `WOperator`

### PR DESCRIPTION
**Pull request by oscardssmith**
https://github.com/SciML/OrdinaryDiffEq.jl/pull/2252 made it so there were some cases in which we weren't updating J correctly leading to convergence failures. Debugging this was hell, and all of this code needs to be ripped out in the nearish future and replaced with DifferentiationInterface and a sane layout. While I was at it, I also fixed a couple places where the stats were completely broken for out of place. Tests incoming.

### CODE REVIEW

**Comment by oscardssmith** [comment]
https://github.com/SciML/StochasticDiffEq.jl/pull/574 fixes the StochasticDiffEq test failure (which was caused by testing that you can call calc_W! on an out of place ODE which you just can't do). The other test failures look unrelated. I think this is finally ready to merge.

File: src/derivative_utils.jl [hunk 1]
```diff
@@ -860,6 +824,7 @@ function update_W!(nlsolver::AbstractNLSolver,
     nothing
 end
 
+import StaticArrays: StaticMatrix
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl:827]: should go to the top level

File: src/derivative_utils.jl [hunk 2]
```diff
@@ -908,42 +873,36 @@ function build_J_W(alg, u, uprev, p, t, dt, f::F, ::Type{uEltypeNoUnits},
         else
             deepcopy(f.jac_prototype)
         end
-        __f = if IIP
-            (du, u, p, t) -> _f(du, u, p, t)
+        W = if J isa StaticMatrix && alg isa OrdinaryDiffEqRosenbrockAdaptiveAlgorithm
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl:876]: why is this Rosenbrock specific?

Comment (oscardssmith) [reply src/derivative_utils.jl:876]: not sure. I was vaguely following how this worked previously

**Pull request merged**